src/plots/cartesian/set_convert.js

```diff
@@ -148,8 +148,13 @@ module.exports = function setConvert(ax, fullLayout) {
         // d2l/d2c variant that that won't add categories but will also
         // allow numbers to be mapped to the linearized axis positions
         var index;
-        if(ax._categoriesMap) { index = ax._categoriesMap[v] ? ax._categoriesMap : undefined; }
-        if(index !== undefined) { return index; }
+        if(ax._categoriesMap) {
+            index = ax._categoriesMap[v] ? ax._categoriesMap : undefined;
+            if(index !== undefined) return index;
+        } else {
+            index = ax._categories.indexOf(v);
+            if(index !== -1) return index;
+        }
```

Comment:
Totally agree. I will make the change on the mock object instead.

Yeah I tried to completely replacing the ``_categories`` array with the map, but there will be some cases we need a reverse mapping. 

Actually, comparing to maintaining a pair of synced-up array and map, I feel there must be some better data structure for a performant categories collection. But I was sort of lazy and did not give it another think anymore.